What is logistic regression?
- used to predict the probibality of an outcome
- can be binary
- supervised learning method
- must provide a dataset that already contains the outcomes to train the model



# Loan Approval Prediction
- automate loan eligibilty
- identify customer whose loan will be approved

Solution Steps: 

- Imprt libraries and read data
- Indetify and Deal with missing values
- Creat Dummy Variables
- Normalize the Data
- select relevant columns
- split dataset in training and test datasets
- train and evaluate the model

In [1]:
#Import libraries
import pandas as pd

In [3]:
#Read the data and create a copy
LoanData = pd.read_csv('01Exercise1.csv')

LoanPrep = LoanData.copy()
LoanPrep.head()

gender married   ch  income  loanamt status
0   Male      No  1.0    5849      NaN      Y
1   Male     Yes  1.0    4583    128.0      N
2   Male     Yes  1.0    3000     66.0      Y
3   Male     Yes  1.0    2583    120.0      Y
4   Male      No  1.0    6000    141.0      Y

In [6]:
#Identify the missing data
LoanPrep.isnull().sum(axis = 0)

gender     13
married     3
ch         50
income      0
loanamt    22
status      0
dtype: int64

In [8]:
#Drop the rows with missing values
LoanPrep = LoanPrep.dropna()
LoanPrep.isnull().sum(axis = 0)

gender     0
married    0
ch         0
income     0
loanamt    0
status     0
dtype: int64

In [9]:
#Drop irrevelent column of gender
LoanPrep = LoanPrep.drop(['gender'], axis = 1)

In [16]:
#create dummy variables
LoanPrep = pd.get_dummies(LoanPrep, drop_first = True)

In [17]:
#normalize the data for loanamt and income
from sklearn.preprocessing import StandardScaler
scalar_ = StandardScaler()
LoanPrep['income'] = scalar_.fit_transform(LoanPrep[['income']])
LoanPrep['loanamt'] = scalar_.fit_transform(LoanPrep[['loanamt']])
LoanPrep.head()

ch    income   loanamt  married_Yes  status_Y
1  1.0 -0.128073 -0.194250            1         0
2  1.0 -0.392077 -0.971015            1         1
3  1.0 -0.461621 -0.294478            1         1
4  1.0  0.108246 -0.031380            0         1
5  1.0  0.011017  1.547205            1         1

In [41]:
#Create the X (Independent) and Y (Dependent dataframes)
Y = LoanPrep[['status_Y']]
X = LoanPrep.drop(['status_Y'], axis = 1)

#Split the X and Y dataset into training and testing set
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.3, random_state = 1234, stratify = Y)


In [42]:
#Build the logistic regression model
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression()

lr.fit(X_train, Y_train)

C:\Users\mkmos\anaconda3\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


LogisticRegression()

In [43]:
#Use "lr.predict" to predict if the loans will be approved
Y_predict = lr.predict(X_test)
Y_predict

array([0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1,
       0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0,
       1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1,
       1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1,
       1, 1, 1, 1, 1], dtype=uint8)

In [45]:
#Check if the model is accurate using a confusion matrix
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(Y_test, Y_predict)
print(cm)

#check the score of our model
lr.score(X_test,Y_test)

[[ 29  20]
 [  2 108]]


0.8616352201257862

According to the confusion matrix our model accurately predicted 29 applicatoins that the loans will be rejected and 108 that the loans would be approved. We incorrectly predicted the outcome of 22 applications. This model is mostly accurate as it predicted 138 out of 159 of the applications or 86% accurate. 